File: ds/trees/level_order_traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# 1. My solution 

import collections
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        l = 1
        q = [(root, l)] #[(node, 1)]
        r = []
        d = collections.defaultdict(list)
        while q:
            c, l = q.pop(0)
            
            if not c:
                continue
            d[l].append(c.val) #dict to keep track of level
            
            q.append((c.left, l+1))
            q.append((c.right, l+1))
                    
        for n in d.values():
            r.append(n)
        
        return r

# BFS with queue
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
    
        res = []
        def bfs(root):
            nonlocal res
            q = [root]
            res = []

            while q:
                t = []
                level_num = len(q)

                for _ in range(level_num):
                    node = q.pop(0)
                    t.append(node.val)
                    
                    if node.left:
                        q.append(node.left)
                    if node.right:
                        q.append(node.right)

                res.append(t)

        bfs(root)

        return res            

File: java/level_order_traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        if (root == null){
            return ans;
        }
        Queue<TreeNode> q = new ArrayDeque<>();
        
        q.add(root);
        while (q.size() > 0){
            int n = q.size();
            ArrayList<Integer> curLevel = new ArrayList<>();

            for (int i = 0; i < n; i++){
                TreeNode cur = q.remove();
                curLevel.add(cur.val);
                if (cur.left != null){
                    q.add(cur.left);
                }
                if (cur.right != null){
                    q.add(cur.right);
                }
            }
            ans.add(curLevel);
        }

        return ans;
    }
}